sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/estimation/FilterEstimation.scala

```diff
@@ -0,0 +1,358 @@
+        evaluateBinary(op, planStat, ar, l, update)
+      case op@GreaterThan(l: Literal, ExtractAttr(ar)) =>
+        evaluateBinary(LessThan(ar, l), planStat, ar, l, update)
+
+      case op@GreaterThanOrEqual(ExtractAttr(ar), l: Literal) =>
+        evaluateBinary(op, planStat, ar, l, update)
+      case op@GreaterThanOrEqual(l: Literal, ExtractAttr(ar)) =>
+        evaluateBinary(LessThanOrEqual(ar, l), planStat, ar, l, update)
+
+      /** EqualTo does not care about the order */
+      case op@EqualTo(ExtractAttr(ar), l: Literal) =>
+        evaluateBinary(op, planStat, ar, l, update)
+      case op@EqualTo(l: Literal, ExtractAttr(ar)) =>
+        evaluateBinary(op, planStat, ar, l, update)
+
+      case In(ExtractAttr(ar), expList) if !expList.exists(!_.isInstanceOf[Literal]) =>
+        /**
+         * Expression [In (value, seq[Literal])] will be replaced with optimized version
+         * [InSet (value, HashSet[Literal])] in Optimizer, but only for list.size > 10.
+         * Here we convert In into InSet anyway, because they share the same processing logic.
+         */
+        val hSet = expList.map(e => e.eval())
+        evaluateInSet(planStat, ar, HashSet() ++ hSet, update)
+
+      case InSet(ExtractAttr(ar), set) =>
+        evaluateInSet(planStat, ar, set, update)
+
+      case Like(_, _) | Contains(_, _) | StartsWith(_, _) | EndsWith(_, _) =>
+        evaluateLike(condition, planStat, update)
+
+      // TODO: it's difficult to estimate IsNull after outer joins
+      // case IsNull(ExtractAttr(ar)) =>
+      // evaluateIsNull(planStat, ar, update)
+      case IsNotNull(ExtractAttr(ar)) =>
+        evaluateIsNotNull(planStat, ar, update)
+
+      case _ =>
+        logDebug("[CBO] Unsupported filter condition: " + condition)
+        notSupported = true
+        1.0
+    }
+    if (notSupported) {
+      1.0
+    } else if (isNot) {
+      1.0 - percent
+    } else {
+      percent
+    }
+  }
+
+  def evaluateIsNotNull(
+      planStat: Statistics,
+      attrRef: AttributeReference,
+      update: Boolean)
+    : Double = {
+    if (!planStat.colStats.contains(attrRef.name)) {
+      logInfo("[CBO] No statistics for " + attrRef)
+      return 1.0
+    }
+    val aColStat = planStat.colStats(attrRef.name)
+    val percent: BigDecimal = planStat.rowCount match {
+      case Some(r) =>
+        if (r == 0) 0.asInstanceOf[BigDecimal]
+        else BigDecimal(aColStat.nullCount) / BigDecimal(r)
+      case None => 0.asInstanceOf[BigDecimal]
+    }
+    if (update) {
+      val newStats = aColStat.copy(nullCount = 0)
+      mutableColStats += (attrRef.name -> newStats)
+    }
+    1.0 - percent.toDouble
+  }
+
+  /** This method evaluates binary comparison operators such as =, <, <=, >, >= */
+  def evaluateBinary(
+      op: BinaryComparison,
+      planStat: Statistics,
+      attrRef: AttributeReference,
+      literal: Literal,
+      update: Boolean)
+    : Double = {
+    if (!planStat.colStats.contains(attrRef.name)) {

```

Comment:
We should put all these checks in the case clause: `case xxx if colStats.contains(attr.name)`